velox/exec/TableScan.cpp

```diff
@@ -57,8 +58,26 @@ RowVectorPtr TableScan::getOutput() {
     return nullptr;
   }
 
+  const auto startTimeMs = getCurrentTimeMs();
   for (;;) {
     if (needNewSplit_) {
+      // Check if our task needs us to yield or we've been running for too long
+      // w/o producing a result. In this case we setup a fake blocking reason
+      // and one already fulfilled future.
+      if (this->driverCtx_->task->shouldStop() != StopReason::kNone or
```

Comment:
@aditi-pandit 

I tried to explain this in the description:

This is done to avoid long loops inside the operator for corner case queries, which have highly filtering scan and return no rows for many splits.
This allows operator to help Driver exist the thread then the Task has requested it and also would eliminate false positives for stuck operator calls.

To elaborate on it:

For some queries (corner case) TableScan::getOutput() can spin for a significant amount of time because the pushed down filter would filter out all rows and we won't get results to break the internal loop and return from the TableScan::getOutput().

For this reason, before grabbing the next split we check if the Task wants us to go off the thread or that we have been running for "too long" (table_scan_getoutput_time_limit_ms).
This has nothing to do with the query execution limit, which is taken care of by the coordinator.

If you look at the Driver's runInternal loop - it exist when Task requests (yield (200ms currently), stop, cancel or such). But TableScan operator is the only one that can 'hog' the thread for a long time w/o exiting.

It is also needed to avoid having false positives in the "stuck operators" report (https://github.com/facebookincubator/velox/pull/6988), which we plan to use to report operators that are likely stuck.

This is in collaboration with @oerling  - he pointed out at the possibility to spend a lot of time inside TableScan::getOutput().

Reply:
Thanks @spershin. This is helpful.